contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle/checkstyle.py

```diff
@@ -0,0 +1,192 @@
+
+  @memoized_property
+  def checker_target(self):
+    self.context.resolve(self._CHECKER_ADDRESS_SPEC)
+    return self.context.build_graph.get_target(Address.parse(self._CHECKER_ADDRESS_SPEC))
+
+  def checker_pex(self, interpreter):
+    # TODO(John Sirois): Formalize in pants.base?
+    pants_dev_mode = os.environ.get('PANTS_DEV')
+
+    if pants_dev_mode:
+      checker_id = self.checker_target.transitive_invalidation_hash()
+    else:
+      checker_id = hash_all([self._CHECKER_REQ])
+
+    pex_path = os.path.join(self.workdir, 'checker', checker_id, str(interpreter.identity))
+    if not os.path.exists(pex_path):
+      with self.context.new_workunit(name='build-checker'):
+        with safe_concurrent_creation(pex_path) as chroot:
+          builder = PEXBuilder(path=chroot, interpreter=interpreter)
+
+          if pants_dev_mode:
+            pex_build_util.dump_sources(builder, tgt=self.checker_target, log=self.context.log)
+            req_libs = [tgt for tgt in self.checker_target.closure()
+                        if isinstance(tgt, PythonRequirementLibrary)]
+            pex_build_util.dump_requirement_libs(builder,
+                                                 interpreter=interpreter,
+                                                 req_libs=req_libs,
+                                                 log=self.context.log)
+          else:
+            pex_build_util.dump_requirements(builder,
+                                             interpreter=interpreter,
+                                             reqs=[PythonRequirement(self._CHECKER_REQ)],
+                                             log=self.context.log)
+          builder.set_entry_point(self._CHECKER_ENTRYPOINT)
+          builder.freeze()
+
+    return PEX(pex_path, interpreter=interpreter)
+
+  def checkstyle(self, interpreter, sources):
+    """Iterate over sources and run checker on each file.
+
+    Files can be suppressed with a --suppress option which takes an xml file containing
+    file paths that have exceptions and the plugins they need to ignore.
+
+    :param sources: iterable containing source file names.
+    :return: (int) number of failures
+    """
+    checker = self.checker_pex(interpreter)
+
+    args = [
+      '--root-dir={}'.format(get_buildroot()),
+      '--severity={}'.format(self.get_options().severity),
+    ]
+    if self.get_options().suppress:
+      args.append('--suppress={}'.format(self.get_options().suppress))
+    if self.get_options().strict:
+      args.append('--strict')
+
+    with temporary_file(binary_mode=False) as argfile:
+      for plugin_subsystem in _PLUGIN_SUBSYSTEMS:
+        options_blob = plugin_subsystem.global_instance().options_blob()
+        if options_blob:
+          argfile.write('--{}-options={}\n'.format(plugin_subsystem.plugin_type().name(),
+                                                   options_blob))
+      argfile.write('\n'.join(sources))
+      argfile.close()
```

Comment:
At least a `flush` is. We're about to hand the now written temp file off to a subprocess to read in. Here the `with` statement nets us an `unlink` (of course a `close` happens too, but the mental model should be that the resource is the file not the file handle as it would be using the `open` context manager).